src/__private_api.rs

```diff
@@ -1,55 +1,87 @@
 //! WARNING: this is not part of the crate's public API and is subject to change at any time
 
+use self::sealed::KVs;
 use crate::{Level, Metadata, Record};
 use std::fmt::Arguments;
-pub use std::option::Option;
 pub use std::{file, format_args, line, module_path, stringify};
 
+#[cfg(feature = "kv_unstable")]
+pub type Value<'a> = dyn crate::kv::value::ToValue + 'a;
+
 #[cfg(not(feature = "kv_unstable"))]
-pub fn log(
+pub type Value<'a> = str;
+
+mod sealed {
+    /// Types for the `kv` argument.
+    pub trait KVs<'a> {
+        fn into_kvs(self) -> Option<&'a [(&'a str, &'a super::Value<'a>)]>;
+    }
+}
+
+// Types for the `kv` argument.
+
+impl<'a> KVs<'a> for &'a [(&'a str, &'a Value<'a>)] {
+    #[inline]
+    fn into_kvs(self) -> Option<&'a [(&'a str, &'a Value<'a>)]> {
+        Some(self)
+    }
+}
+
+impl<'a> KVs<'a> for () {
+    #[inline]
+    fn into_kvs(self) -> Option<&'a [(&'a str, &'a Value<'a>)]> {
+        None
+    }
+}
+
+// Log implementation.
+
+fn log_impl(
     args: Arguments,
     level: Level,
     &(target, module_path, file): &(&str, &'static str, &'static str),
     line: u32,
-    kvs: Option<&[(&str, &str)]>,
+    kvs: Option<&[(&str, &Value)]>,
 ) {
+    #[cfg(not(feature = "kv_unstable"))]
     if kvs.is_some() {
         panic!(
             "key-value support is experimental and must be enabled using the `kv_unstable` feature"
         )
     }
 
-    crate::logger().log(
-        &Record::builder()
-            .args(args)
-            .level(level)
-            .target(target)
-            .module_path_static(Some(module_path))
-            .file_static(Some(file))
-            .line(Some(line))
-            .build(),
-    );
+    let mut builder = Record::builder();
+
+    builder
+        .args(args)
+        .level(level)
+        .target(target)
+        .module_path_static(Some(module_path))
+        .file_static(Some(file))
+        .line(Some(line));
+
+    #[cfg(feature = "kv_unstable")]
+    builder.key_values(&kvs);
+
+    crate::logger().log(&builder.build());
 }
 
-#[cfg(feature = "kv_unstable")]
-pub fn log(
+pub fn log<'a, K>(
     args: Arguments,
     level: Level,
-    &(target, module_path, file): &(&str, &'static str, &'static str),
+    target_module_path_and_file: &(&str, &'static str, &'static str),

```

Comment:
Besides, does this change affect significantly? Or just some identical refactor.